### PR TITLE
[mypy] Enable following imports for some directories

### PR DESCRIPTION
**Pull request by DarkLight1337**
The `follow_imports` option has been enabled for the following files and directories:

- `vllm/*.py` (originally failed `mypy`, made passing in this PR)
- `vllm/adapter_commons` (originally failed `mypy`, made passing in this PR)
- `vllm/assets`
- `vllm/inputs`
- `vllm/logging`
- `vllm/multimodal`
- `vllm/platforms`
- `vllm/server`
- `vllm/transformers_utils` (originally failed `mypy`, made passing in this PR)
- `vllm/triton_utils` (originally failed `mypy`, made passing in this PR)
- `vllm/usage`

Future code modifying those files and directories will thus be subject to stricter type checks - another step towards fulfilling #3680.

Note that the config in `pyproject.toml` has been updated such that `follow_imports` is now enabled by default, but only checks the above directories. After updating another directory to pass with `follow_imports` enabled, please add a corresponding entry in `pyproject.toml` then in turn remove it from `format.sh` and `mypy.yaml`.

### CODE REVIEW

File: format.sh [hunk 1]
```diff
@@ -96,23 +96,19 @@ echo 'vLLM yapf: Done'
...
-mypy vllm/logging --config-file pyproject.toml
-mypy vllm/lora --config-file pyproject.toml
-mypy vllm/model_executor  --config-file pyproject.toml
-mypy vllm/multimodal --config-file pyproject.toml
-mypy vllm/prompt_adapter --config-file pyproject.toml
-mypy vllm/spec_decode --config-file pyproject.toml
-mypy vllm/transformers_utils --config-file pyproject.toml
-mypy vllm/usage --config-file pyproject.toml
-mypy vllm/worker --config-file pyproject.toml
+mypy tests --follow-imports skip
```
Comment (rkooo567) [inline format.sh:99]: the new formatter and github action doesn't pass --config-file, is this intended?

File: vllm/utils.py [hunk 2]
```diff
@@ -94,8 +94,10 @@ def __contains__(self, key: Hashable) -> bool:
     def __len__(self) -> int:
         return len(self.cache)
 
-    def __getitem__(self, key: Hashable) -> Optional[T]:
-        return self.get(key)
+    def __getitem__(self, key: Hashable) -> T:
```
Comment (rkooo567) [inline vllm/utils.py:97]: is this related to this PR?

File: vllm/utils.py [hunk 3]
```diff
@@ -109,8 +111,9 @@ def touch(self, key: Hashable) -> None:
     def get(self,
             key: Hashable,
             default_value: Optional[T] = None) -> Optional[T]:
+        value: Optional[T]
```
Comment (rkooo567) [inline vllm/utils.py:114]: this doesn't seem to be optional anymore this way?

Comment (rkooo567) [reply vllm/utils.py:114]: (iiuc if there's no item, it should raise an exception?)

**Review by rkooo567 — APPROVED** [review]

File: vllm/utils.py (hunk 3 above)
Comment (DarkLight1337) [reply vllm/utils.py:114]: Yes, that is the intention so it works more like a Python dictionary.

File: vllm/utils.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/utils.py:97]: See below.

File: format.sh (hunk 1 above)
Comment (DarkLight1337) [reply format.sh:99]: Yes. The TOML file is already being used by default so there is no need to specify it explicitly
